doc/sphinx/proof-engine/ltac.rst

```diff
@@ -193,16 +134,114 @@ evaluation yields either a term, an integer or a tactic. Intermediate
 results can be terms or integers but the final result must be a tactic
 which is then applied to the focused goals.
```

Comment:
GitHub won't let me comment on the full paragraph, but I'd suggest something like:
```rst
|Ltac| has two notions of tactics: tactic *expressions* and tactic *values*.
A tactic *expression* is anything the user types in.
A tactic *value* is either a tactic that is ready to be run on the goal (such as the :tacn:`constructor` tactic),
a tactic function (such as `ltac:(fun _ => constructor)`), unit (denoted `()`), or anything in the table of :ref:`Syntactic values`,
such as integers, strings, constrs, etc.
|Ltac| evaluation has two phases: expression evaluation phase, when tactic expressions are evaluated into tactic values,
and the runtime phase, when tactic values are run in the context of goals.
During the expression evaluation phase, the following things happen:
- If the top-level expression is of the form `let name args := body in rest`, then `body` is evaluated in the context of `args` and bound to `name`, after which `rest` is evaluated
- if the top-level expression is any function application of the form `f args`, then `f` is expression-evaluated, then `args` are expression-evaluated from left to right, and `f` is applied to `args`, and the result is expression-evaluated
- if the top-level expression is a `match`, `lazymatch`, or `multimatch`, then the `match` is expression-evaluated *and* the selected branch of the match is *both* expression-evaluated *and run as a tactic on the current goal(s)*.
- If the top-level expression is anything else, then it is returned as-is as a tactic value.
```

This might deserve a couple of examples, and can probably be polished, but it's a first stab.

Reply:
Initial thoughts:
- "tactic value" as you use it here is confusing.  The obvious meaning would be "a value of type *tactic*", while you describe it as including integers, strings, etc.  It also clashes with the error message "Expression does not evaluate to a tactic." from `integer:(1).`  Maybe best just to refer to these as "values".

- For a given complex Ltac expression, is all of the evaluation phase done before any of the runtime phase is done?  (I'm guessing no)  Or do the phases apply recursively to individual constructs?  Seems like a let inside a match action might have to be recomputed if there are multiple matches for the branch.

- Which of the following, if any, are only part of the evaluation phase?  I think we'd want to be quite clear about this.
  - substitution of matched values into a tactic
  - getting the value generated by "fresh"
  - getting the term generated by the "context" tactic

Reply:
> * "tactic value" as you use it here is confusing.  The obvious meaning would be "a value of type _tactic_", while you describe it as including integers, strings, etc.  It also clashes with the error message "Expression does not evaluate to a tactic." from `integer:(1).`  Maybe best just to refer to these as "values".

Yes, "values" sounds better.  Or maybe "Ltac values", to make it clear that we're talking about values in the context of Ltac, not in, e.g., the context of metatheory of CiC terms.

> For a given complex Ltac expression, is all of the evaluation phase done before any of the runtime phase is done? (I'm guessing no) Or do the phases apply recursively to individual constructs?

The phases apply recursively.  For example, the *runtime* behavior of sequencing (`;`) is that to *run* `a; b`, you first expression-evaluate `a` to a value, then you *run* that value, then you expression-evaluate `b` to a value, and then you run that value.  This ordering can be observed with the example:
```coq
Goal True.
  (let __ := match goal with _ => idtac 0 end in idtac 1);
    (let __ := match goal with _ => idtac 2 end in idtac 3).
  (* 0 1 2 3 *)
```
This is why prefixing a tactic `idtac; ` delays the expression-evaluation of that tactic.  Note that this is in contrast to function application and tactic notation evaluation, which evaluate all of their arguments first:
```coq
Tactic Notation "seq" tactic3(a) "then" tactic3(b) := a; b.

Goal True.
  seq (let __ := match goal with _ => idtac 0 end in idtac 1)
  then (let __ := match goal with _ => idtac 2 end in idtac 3).
  (* 0 2 1 3 *)
```

> Seems like a let inside a match action might have to be recomputed if there are multiple matches for a branch.

All `match` constructs force eager *running* of their branches, allowing users to use `match` to "inject" runtime behavior into expression-evaluation phase, as I do in the above examples.  So I'm not sure what you mean by "recomputed".

> Which of the following, if any, are only part of the evaluation phase? I think we'd want to be quite clear about this.
>
> - substitution of matched values into a tactic

What do you mean by this?  I think perhaps it's better to model Ltac as being an imperative language with a context, and variables are looked up in the context as needed, rather than thinking of substituting values into a "tactic AST".  So perhaps there are properly three phases: (1) parsing/interpretation of a user-written AST into an unevaluated expression, (2) expression evaluation, (3) running of tactics on goals.  (2) and (3) are interwoven with each other, but not with (1).

> - getting the value generated by "fresh"

Yes, I guess I forgot about this, things like `fresh` and `numgoals` are run/evaluated entirely during expression-evaluation.

> - getting the term generated by the "context" tactic

I wouldn't call `context` a tactic, but a syntactic construct, but, indeed, this too occurs during expression-evaluation.  So does running `eval redexpr in term`.

I believe the full list is: the two non-`info` things in `tactic:binder_tactic` (`fun` and `let`), everything in `tactic:tactic_arg` (which includes `eval ...`, `fresh`, `type_term`, `numgoals`, among others), `match`, `lazymatch`, `multimatch`, substitution/context-lookup of things in `prim:reference` (under `tactic:tactic_expr`), the `LIST0 tactic_arg_compat` (under `tactic:tactic_expr`) which I believe is used for function application / function calling, `( ... )` which is just used for grouping (at level 0 in `tactic:tactic_expr`), `tactic_atom` (`()` and integers and qualids), and the things in `tactic_arg_compat` (which seem to be `tactic_arg`, `Constr.constr`, and `()`).

Reply:
Maybe we're taking this too far?  Aside from `assert_*`, when is the distinction between evaluation and runtime really important to the user?  Clearly it affects your `"seq"` example, but is it a central point or a lesser detail?  Might be easier to show that example.  Probably safest to suggest users experiment with it to understand the order (i.e. it's not always what you expect) rather than trying to be authoritative.

The text so far doesn't cover a number of Ltac constructs.  It's complicated stating it correctly, e.g. for "If the top-level expression is of the form `let name args := body in rest`, then `body` is evaluated in the context of `args` and bound to `name`, after which `rest` is evaluated", `body` is also run (maybe "execute"/"execution" is clearer than "runs"/"runtime"?).

Reply:
>  when is the distinction between evaluation and runtime really important to the user?

It's absolutely essential when writing a tactic that should both have side effects and return a constr, integer, ident, etc, and you don't want to cps-convert the tactic.  Consider:
```coq
Ltac cache_term val id :=
  let H := fresh in
  let id := fresh id in
  let ty := type of val in
  let __ := lazymatch goal with
            | _ => simple refine (let H : ty := _ in _);
                   [ transparent_abstract exact val using id | ]
            end in
  let ret := (eval cbv delta [H] in H) in
  let __ := lazymatch goal with _ => clear H end in
  ret.

Goal True.
  let v := cache_term I val in
  idtac v. (* val *)
  Print val. (* val = I
     : True *)
```

> maybe "execute"/"execution" is clearer than "runs"/"runtime"?

Sure.

Reply:
What is cps-convert?  And in which part of the example is the distinction between evaluation and runtime essential?  On `ret.`?  BTW, AFAIK we don't mention function return values anywhere in the chapter.

Seems like some constructs don't play well with integers.  Is this worth mentioning at all?

```
Goal True /\ False.
let x := tryif idtac then 1 else 2 in idtac x.   (* <tactic closure> *)
let x := match goal with |- _ => 3 | |- False => 4 end in idtac x.  (* 3 *)
let x := split; match goal with |- True => 3 | |- False => 4 end in idtac x.  (* <tactic closure> *)
```

It seems to me that the important point for users in your proposed text is the part about match/lazymatch/multimatch, which would be unexpected, and that the rest is a very standard way to evaluate expressions.  In which case I would only discuss that single point.  WDYT?  Am I mistaken?

Reply:
> Seems like some constructs don't play well with integers. Is this worth mentioning at all?

This is exactly the difference between expression-evaluation and runtime.  The constructs that play well with integers (also: constrs, uconstrs, etc) are exactly the ones that are evaluated during expression-evaluation.

> It seems to me that the important point for users in your proposed text is the part about match/lazymatch/multimatch, which would be unexpected, and that the rest is a very standard way to evaluate expressions. In which case I would only discuss that single point. WDYT? Am I mistaken?

I think this is one of the points, and the other point to discuss is just that there *are* two separate phases in the first place.